### PR TITLE
Tighten validation on the qosClass field of pod status

### LINKED ISSUE #126662: QosClass of pod status shouldn't be changeable
### What happened?

According to this [code](https://github.com/kubernetes/kubernetes/blob/bbe8ca8b2ab14992389bc67e3bcfa209adcb13d4/pkg/apis/core/validation/validation.go#L4996), the QosClass of pod status is immutable. 

This [function ](https://github.com/kubernetes/kubernetes/blob/40b604e374144351eac463e7077fdb1903f5a033/pkg/apis/core/helper/qos/qos.go#L48
)also shows that QosClass is computed from the resources instead of being provided by users. 

However, we can use rest api or "kubectl edit pod <podName> --subresource='status'" to change it. 

### What did you expect to happen?

The request to change QoSClass of pod status should receive a validation error to indicate it's immutable. 

### How can we reproduce it (as minimally and precisely as possible)?

Use "kubectl edit pod <podName> --subresource='status'" to change the qosClass to a different value. It will succeed. 

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: v1.29.4
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.29.4
```

</details>

### Cloud provider

<details>
Azure
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by carlory**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

According to this [code](https://github.com/kubernetes/kubernetes/blob/bbe8ca8b2ab14992389bc67e3bcfa209adcb13d4/pkg/apis/core/validation/validation.go#L4996), the QosClass of pod status is immutable.

This [function ](https://github.com/kubernetes/kubernetes/blob/40b604e374144351eac463e7077fdb1903f5a033/pkg/apis/core/helper/qos/qos.go#L48)also shows that QosClass is computed from the resources instead of being provided by users.

However, we can use rest api or "kubectl edit pod --subresource='status'" to change it.

#### Which issue(s) this PR fixes:

Fixes #126662

#### Does this PR introduce a user-facing change?

```release-note
Tighten validation on the qosClass field of pod status. This field is immutable but it would be populated with the old status by kube-apiserver if it is unset in the new status when updating this field via the status subsource.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by ffromani** [comment]
/triage accepted
/priority backlog

priority set based on the fact this is a very old issue we endured so far
I agree users should not be allowed to set QoSClass. This field is owned and set by the system

**Comment by liggitt** [comment]
As long as ComputePodQOS *only* looks at spec fields, I think this makes sense.

If ComputePodQOS *ever* starts to look at status fields that can change, I think this might get us into trouble.

@tallclair, what is the plan for ensuring in-place resource changes can or cannot change PodQOS?

**Comment by carlory** [comment]
e2e is broken, it will be fixed in https://github.com/kubernetes/kubernetes/pull/128083

**Comment by liggitt** [comment]
I'd like a first pass by @tallclair on the question I asked

**Comment by carlory** [comment]
Hi @liggitt, what do you think about @SergeyKanzhelev‘s https://github.com/kubernetes/kubernetes/issues/126662#issuecomment-2414412835 on this issue?

**Comment by liggitt** [comment]
routing API approval to @thockin as well to make sure this is coherent with https://github.com/kubernetes/kubernetes/pull/128266/files#r1815329831

**Comment by SergeyKanzhelev** [comment]
/hold

even if we decide to go ahead with this, we cannot merge it for _at least_ two releases since old kubelet is resetting the QoS on pod admission failure: https://github.com/kubernetes/kubernetes/pull/128083

**Comment by thockin** [comment]
Would we want to allow this to merge with a special-case for setting it to `""`, perhaps also check when we know it is being rejected by kubelet?

**Comment by thockin** [comment]
I think the plan, even with in-place updates, is to never change this (even if updates to spec might make it LOOK LIKE a different QoS (which is currently disallowed))

**Comment by SergeyKanzhelev** [comment]
> Would we want to allow this to merge with a special-case for setting it to `""`, perhaps also check when we know it is being rejected by kubelet?

it is a very good idea. I will be ok with this. Even if it will disallow some updates, it will make more good than bad by keeping the QoS constant. And we do not need to wait for two releases.

**Comment by carlory** [comment]
> Would we want to allow this to merge with a special-case for setting it to ""

Updated. @SergeyKanzhelev An empty string is allowed when updating a Pod's status, so other changes in the intergation and e2e tests can be reverted, so I want to confirm wether I shoud remove these changes of all test files.

**Comment by carlory** [comment]
@SergeyKanzhelev Another question is that if the pr is merged, do we need to revert the https://github.com/kubernetes/kubernetes/pull/128083?

**Comment by SergeyKanzhelev** [comment]
> do we need to revert the https://github.com/kubernetes/kubernetes/pull/128083?

No, let's keep it. Removing the QoS just doesn't make sense.

File: pkg/apis/core/validation/validation.go [hunk 1]
```diff
@@ -5373,6 +5373,11 @@ func ValidatePodStatusUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions
 		}
 	}
 
+	// For backward compatibility, we allow the qosClass to be set to an empty string.
+	if newPod.Status.QOSClass != oldPod.Status.QOSClass && len(newPod.Status.QOSClass) > 0 {
```
Comment (SergeyKanzhelev) [inline pkg/apis/core/validation/validation.go:5377]: Should we also allow to set a new QoS class when the old QoS class was empty?

File: pkg/apis/core/validation/validation.go [hunk 2]
```diff
@@ -5373,6 +5373,11 @@ func ValidatePodStatusUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions
...
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("qosClass"), "must be original value or empty string"))
```
Comment (SergeyKanzhelev) [inline pkg/apis/core/validation/validation.go:5378]: insert the original value in the error message

File: pkg/apis/core/validation/validation_test.go [hunk 3]
```diff
@@ -14454,6 +14454,28 @@ func TestValidatePodStatusUpdate(t *testing.T) {
 		),
 		"",
 		"restartable init container can restart if RestartPolicyAlways",
+	}, {
```
Comment (SergeyKanzhelev) [inline pkg/apis/core/validation/validation_test.go:14457]: add a test case that we allow to set to the same value

File: test/integration/auth/node_test.go [hunk 4]
```diff
@@ -254,7 +254,7 @@ func TestNodeAuthorizer(t *testing.T) {
 			startTime := metav1.NewTime(time.Now())
 			_, err := client.CoreV1().Pods("ns").UpdateStatus(context.TODO(), &corev1.Pod{
 				ObjectMeta: metav1.ObjectMeta{Name: "node2normalpod"},
-				Status:     corev1.PodStatus{StartTime: &startTime},
+				Status:     corev1.PodStatus{StartTime: &startTime, QOSClass: corev1.PodQOSBestEffort},
```
Comment (SergeyKanzhelev) [inline test/integration/auth/node_test.go:257]: let's not mix in all other changes into this PR. If we want to improve all these tests, we can do it in separate PR. Let's concentrate this PR to an API validation change

File: pkg/apis/core/validation/validation.go (hunk 1 above)
Comment (carlory) [reply pkg/apis/core/validation/validation.go:5377]: The QOSClass field will be populated by kube-apiserver before a Pod is created. https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/pod/strategy.go#L89

For backward compatibility, the QOSClass field can be overridden by the user or a controller to an empty string. But I don't know whether there is a case where some users or controllers reset the QOSClass field to another value after the field is dropped by others or themselves. cc @thockin 

If we want to the QOSClass field not to be changed by others, we may be able to override the field in the podStatusStrategy's PrepareForUpdate. It may not be a good idea.

Comment (carlory) [reply pkg/apis/core/validation/validation.go:5377]: cc @liggitt

File: pkg/apis/core/validation/validation_test.go (hunk 3 above)
Comment (carlory) [reply pkg/apis/core/validation/validation_test.go:14457]: added.

File: pkg/apis/core/validation/validation.go (hunk 1 above)
Comment (SergeyKanzhelev) [reply pkg/apis/core/validation/validation.go:5377]: Yes, I understand it is not something we will do. It also opens a door to actually changing the QoS class in two changes QoS1 -> empty -> QoS2.

However, consider a scenario: two controllers updating the status of a Pod. One is implemented with the reset to empty (similar to what pod rejection is doing), another is implemented with the keeping the QoS (reading a status and then updating one field). We will end up with the race condition - if first controller executed first, than second one will fail. If second will succeed first, then first will work fine. This situation is not that critical if failed status update is handled correctly. But if it will result in continuous retries, it may cause some problems.

If this example doesn't make sense, and there are other reasons this scenario doesn't make sense, please let me know. I didn't try it myself, but can imagine how it is likely will work.

Comment (thockin) [reply pkg/apis/core/validation/validation.go:5377]: Two possible paths, maybe:

1) If we know that we're compensating for that specific kubelet issue, we could allow changing it to "" ONLY WHEN we are also changing phase to "Failed".  There's no coming back from that, right?  My kingdom for a stat machine diagram.

2) If we know that changing from a value to "" is just for  that kubelet issue, we could manually "fix up" the operation to carry the old value forward.  We do this in a couple places in Service, where we auto-write values to spec.  It's kind of awful but it works.   pkg/registry/core/service/storage/storage.go `patchAllocatedValues()`

```
if needsClusterIP(oldSvc) && needsClusterIP(newSvc) { 
          if newSvc.Spec.ClusterIP == "" { 
              newSvc.Spec.ClusterIP = oldSvc.Spec.ClusterIP
          } 
```

This is called from `beginUpdate` which is installed as the `store.BeginUpdate` hook in `NewREST()`.

If someone tries to change from QoS1 -> QoS2, validation should still fail.

Comment (carlory) [reply pkg/apis/core/validation/validation.go:5377]: After I look at the service code, the second way isn't a better choice. According to the `BeginUpdateFunc` definition, the function should only be used for a "transaction-like" operation. It is obvious that the `BeginUpdateFunc` is not suitable for this case even though it works. the `UpdateStrategy` machinisim is more preferred. It allows us to implement resource-specific behavior during updates.

The qosClass of a Pod is computed according to the PodSpec. I think the best way to resolve this problem is to override it with the old pod‘s status whatever its value from being updated Pod is. Changing or dropping the qosClass of a Pod is not intended to be a user-facing operation, and it is not recommended to allow it to be changed. The kubelet didn't really want to drop this field when it rejected a pod. A reject reason, message, and marking it as failed is all that the kubelet wants to do. we can treat the qosClass as a managed field by the kube-apiserver, any user or external controller can not drop and change it by updating the pod status. No error will be returned because the field will be populated by the kube-apiserver.

We know the InPlacePodVerticalScaling feature disallows to change a pod's qosClass via updating the pod's spec settings. so we don't need to re-compute the pod's qosClass with the new spec. we can just override the qosClass of the pod with the old pod's status.

For example,

```go
func (podStatusStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
	newPod := obj.(*api.Pod)
	oldPod := old.(*api.Pod)
	newPod.Spec = oldPod.Spec
	newPod.DeletionTimestamp = nil

	// don't allow the pods/status endpoint to touch owner references since old kubelets corrupt them in a way
	// that breaks garbage collection
	newPod.OwnerReferences = oldPod.OwnerReferences
        // the Pod QoS is immutable and populated at creation time by the kube-apiserver
        newPod.Status.QOSClass = oldPod.Status.QOSClass
}
```

The benefits of this approach are:
1. No breaking changes for the old kubelet
2. No need to validate the qosClass of the pod in the update validation. If we allow the qosClass to be updated from empty to a string, we have to re-compute the qosClass of the pod with the new spec settings. And then compare the value with the qosClass of the pod's status. 
3. Users or external controllers don't worry about the conflict when updating the pod's status.

What do you think? @SergeyKanzhelev @thockin

Comment (SergeyKanzhelev) [reply pkg/apis/core/validation/validation.go:5377]: I like all the benefits of the override method. Let's do it.

Comment (carlory) [reply pkg/apis/core/validation/validation.go]: @thockin @SergeyKanzhelev  It's only populated when its value is empty. And add a validation to check if it's mutated. Please review it again. Thanks a lot!

**Review by thockin — COMMENTED** [review]
/lgtm
/approve

/hold

@SergeyKanzhelev ?

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm
/unhold
/approve
